### PR TITLE
libcoap: 4.2.1 -> 4.3.0

### PR DESCRIPTION
**Pull request by kmein**
###### Motivation for this change
There is a new version of libcoap which supports TLS (the coaps:// protocol).
This makes it usable for, e.g. the IKEA Trådfri smart lighting zigbee network.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/libcoap/default.nix [hunk 1]
```diff
@@ -1,26 +1,29 @@
...
     automake
     autoconf
     which
     libtool
     pkg-config
-  ];
+  ] ++ lib.optional tls gnutls ++ lib.optionals docs [ doxygen asciidoc ] ;
   preConfigure = "./autogen.sh";
   configureFlags = [
     "--disable-documentation"
```
Comment (ajs124) [inline pkgs/applications/networking/libcoap/default.nix:24]: Does it still build the documentation, if it finds the relevant software?

File: pkgs/applications/networking/libcoap/default.nix [hunk 2]
```diff
@@ -1,26 +1,29 @@
-{ fetchFromGitHub, automake, autoconf, which, pkg-config, libtool, lib, stdenv }:
+{ fetchFromGitHub, automake, autoconf, which, pkg-config, libtool, lib, stdenv, gnutls, asciidoc, doxygen
+, tls ? true
+, docs ? true
```
Comment (ajs124) [inline pkgs/applications/networking/libcoap/default.nix:3]: I think we normally call these something like `withTLS` or `enableTLS`, although I'm not sure there is really any agreed-upon reason for that.

Comment (risicle) [reply pkgs/applications/networking/libcoap/default.nix:3]: Yes, I think "with" is slightly more common.

**Comment by risicle** [comment]
`nixpkgs-review` happy, nixos x86_64 & aarch64

File: pkgs/applications/networking/libcoap/default.nix (hunk 1 above)
Comment (kmein) [reply pkgs/applications/networking/libcoap/default.nix:24]: Good catch! Thank you.

File: pkgs/applications/networking/libcoap/default.nix (hunk 2 above)
Comment (kmein) [reply pkgs/applications/networking/libcoap/default.nix]: I'm using `with` now.

**Comment by siraben** [comment]
Result of `nixpkgs-review pr 152325` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)

**Comment by siraben** [comment]
Builds and runs on x86_64-darwin for me.

**Pull request merged**